fix: Command/Combobox TypeError and Unclickable/Disabled items + Documentation

jhnguyen521 opened the pull request:
Fixes #2944 due to latest cmdk update with breaking changes for the Command component. Until this fix is merged, all new command components that install cmdk@1.00 will be broken and items will not be selectable. People who also have existing components and update to cmdk 1.00 and don’t have a `<CommandList>` will experience crashes upon opening up the component. See issue above for more details.

- Additionally handling the data-disabled property better in all components, should be backwards compatible
^ Happy to not do this for all other components, but I figure it's better to just do this preventatively if it's backwards compatible. Either way, things will break for data-disabled="false" for all existing `data-[disabled]`
- Applied same fix for aria-selected
- Fixing broken examples for command component

kevinmitch14 commented:
This PR handles the proper styling via data attributes + dependancy bump.

https://github.com/shadcn-ui/ui/pull/2626

jhnguyen521 commented:
Fixed crlf line endings to lf. May be worth documenting this in contributing, but pnpm seems to output files with \r\n instead of just \r on windows machines, ignoring tsconfig compiler options

kevinmitch14 commented:
The component specific issues have already been fixed in https://github.com/shadcn-ui/ui/pull/2626 🤔

jhnguyen521 commented:
> The component specific issues have already been fixed in #2626 🤔

After you added the fix from this and the other issue afterwards in a later commit, then yes :) Also, aria is already set in cmdk and we can be relatively certain that will always be there for accessibility reasons, not sure how opinionated we should be about using their aria/data attribute. Personally, I think it's fine to leave as is but what do I know.

Also, this PR updates necessary documentation to handle other breaking changes in cmdk.

jhnguyen521 commented:
Happy to do a version bump in this PR as well.

kevinmitch14 commented:
I am not too sure why you are hinting that I have "copied" your work or something?

With all due respect, the original #2626 PR was opened in January as a follow up from #1980 that was opened in November 2023. I have been monitoring this via an issue opened in the `cmdk` repo itself.

After my initial comment on this PR, I noticed that you made some extra commits, specifically, extra changes that were made in #2626 that had not been made in this PR. I see that these commits have now been wiped out as you cleaned the commit history...

You also commented on my PR which you have since deleted...I don't understand

jhnguyen521 commented:
> After my initial comment on this PR, I noticed that you made some extra commits, specifically, extra changes that were made in https://github.com/shadcn-ui/ui/pull/2626 that had not been made in this PR. I see that these commits have now been wiped out as you cleaned the commit history...

Yes, the change I made were applying my bug fix to the aria-selected attributes, which is the purpose of this PR? Whereas your PR was originally just changing aria-selected -> data-selected.
Huh? Cleaned the commit history? Probably because just did a rebase after some new stuff was merged in, but all commits should be as they have...

> You also commented on my PR which you have since deleted...I don't understand
My comment was on being opinionated about aria-selected vs data-selected, but I realized that I was incorrect about some assumptions I was saying that made my comment irrelevant so I just deleted it :)

CentraXx commented on apps/www/registry/default/ui/select.tsx:
```diff
@@ -118,7 +118,7 @@ const SelectItem = React.forwardRef<
     ref={ref}
     className={cn(
-      "relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50",
+      "relative flex w-full cursor-default select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none focus:bg-accent focus:text-accent-foreground data-[disabled='true']:pointer-events-none data-[disabled='true']:opacity-50",
```
This does not work properly. Please test this...

CentraXx commented on apps/www/registry/default/ui/dropdown-menu.tsx:
```diff
@@ -83,7 +83,7 @@ const DropdownMenuItem = React.forwardRef<
     ref={ref}
     className={cn(
-      "relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled]:pointer-events-none data-[disabled]:opacity-50",
+      "relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none transition-colors focus:bg-accent focus:text-accent-foreground data-[disabled='true']:pointer-events-none data-[disabled='true']:opacity-50",
```
This does not work properly. Please test this...

CentraXx reviewed (commented):
For many components which are not from cmdk the applied changes do not work and cause the disable state to not work anymore.
Means the fix: data-[disabled='true'] does not work for all components.

jhnguyen521 commented:
> For many components which are not from cmdk the applied changes do not work and cause the disable state to not work anymore. Means the fix: data-[disabled='true'] does not work for all components.

Good catch, thanks! I'll probably just revert changes to all other components since I'd probably have to make a custom class in css to get this to work for all while still being in the classNames.

benjamin-guibert commented:
Thank you for the fix!

digitaltim-de commented:
Same Issue on the Example Filter Function, that dont works anymore too.

nopitown commented:
Thanks for the fix!

logan-qiu commented:
@jhnguyen521 thank you! I am pretty new to shadcn and figured it will lead to a crash if I follow the official example in `ui/apps/www/app/(app)/examples/forms/account/account-form.tsx` for the language selector.
I had to wrap CommandGroup with CommandList to avoid the crash. However, none of the selection was selectable.

Hopefully this fix will be merged into the main branch soon! kudos to you.

valascus commented:
When will this be merged?
Thank you

wootra reviewed (approved):
reviewed the code and tested in the code space. 
looks good to me.

kaiguang commented:
Copied the code from the official website and found out it's not working with `cmdk@1.0.0`, maybe we can lock the package version with the install instruction?

<img width="399" alt="Screenshot 2024-05-27 at 21 04 41" src="https://github.com/shadcn-ui/ui/assets/33639861/a490e12c-3b80-46a4-89e8-e2ccde691e04">

shadcn commented:
@jhnguyen521 Any chance you can fix the conflicts here please? Let's merge.

kevinmitch14 commented:
@shadcn check out #2626 which is a follow up of #1980. Both pending for a long time addressing this issue.

jhnguyen521 commented:
> @jhnguyen521 Any chance you can fix the conflicts here please? Let's merge.

Conflicts are fixed :)

ivan2214 reviewed (approved):
LGTM

shadcn commented on apps/www/registry/default/ui/command.tsx:
```diff
@@ -117,7 +117,7 @@ const CommandItem = React.forwardRef<
     ref={ref}
     className={cn(
-      "relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none data-[disabled=true]:pointer-events-none data-[selected=true]:bg-accent data-[selected=true]:text-accent-foreground data-[disabled=true]:opacity-50",
+      "relative flex cursor-default select-none items-center rounded-sm px-2 py-1.5 text-sm outline-none aria-[selected='true']:bg-accent aria-[selected='true']:text-accent-foreground data-[disabled='true']:pointer-events-none data-[disabled='true']:opacity-50",
```
Can we use the new data-selected attribute instead if aria? Thank you.

kotAPI commented:
Any updates on this?

saver711 commented:
WE NEED THIS !!!

shadcn commented:
@kevinmitch14 I merged https://github.com/shadcn-ui/ui/pull/2626. Anything else we're missing? \cc @jhnguyen521

kevinmitch14 commented:
@shadcn it looks like some changes made in #2626 were reverted in https://github.com/shadcn-ui/ui/pull/4181/ by accident.

kevinmitch14 commented:
@shadcn it looks like some changes made in #2626 were reverted in https://github.com/shadcn-ui/ui/pull/4181/ by accident. (`command.tsx`, `model-selector.tsx`, `cmdk` package was reverted back to v0.2.0 files)

shadcn commented:
@kevinmitch14 Thanks. On it.
